### DISCUSSION TITLE
Getting "failed to install" on numerous OpenTelemetry Instrumentations

### DISCUSSION THREAD

**Discussion by 3ygun** [open-telemetry/opentelemetry-ruby#1050] [Q&A] [answered]
Hello,

I followed the [Getting Started](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/) page mapping each direction into my Ruby on Rails application. However, when I boot the rails service I get the following (I'm using customized `c.use`s to avoid noise see the config below):

```text
D, [2021-12-02T19:50:30.196618 #10] DEBUG -- : Upgrading default proxy tracer provider to OpenTelemetry::SDK::Trace::TracerProvider
W, [2021-12-02T19:50:30.196870 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::ActionPack failed to install
W, [2021-12-02T19:50:30.196947 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::ActionView failed to install
W, [2021-12-02T19:50:30.197015 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::ActiveJob failed to install
W, [2021-12-02T19:50:30.197151 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::ActiveModelSerializers failed to install
W, [2021-12-02T19:50:30.197301 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::ActiveRecord failed to install
I, [2021-12-02T19:50:30.198297 #10]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::ConcurrentRuby was successfully installed
I, [2021-12-02T19:50:30.199680 #10]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::Dalli was successfully installed
W, [2021-12-02T19:50:30.199892 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::HttpClient failed to install
I, [2021-12-02T19:50:30.201298 #10]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::Mysql2 was successfully installed
I, [2021-12-02T19:50:30.203433 #10]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::Rack was successfully installed
W, [2021-12-02T19:50:30.203653 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::Rails failed to install
I, [2021-12-02T19:50:30.204584 #10]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::Redis was successfully installed
W, [2021-12-02T19:50:30.204806 #10]  WARN -- : Instrumentation: OpenTelemetry::Instrumentation::Sidekiq failed to install
I, [2021-12-02T19:50:30.204934 #10]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::Sinatra was successfully installed
```

Thus I can see some of the traces (redis, mysql, etc.) in Jaeger but not the ones from the core Rails glue of the application.
I've tried to get more information from the loggers as detailed in https://github.com/open-telemetry/opentelemetry-ruby/discussions/966 (see the `error_handler` below) but I'm having no luck getting logs beyond the:

> WARN -- : Instrumentation: OpenTelemetry::Instrumentation::Rails failed to install

Using rails `5.0.7.2`

My `config/initializers/opentelemetry.rb` is:

```ruby
require 'opentelemetry/sdk'
require 'opentelemetry/exporter/otlp'

OpenTelemetry::SDK.configure do |c|
  updox_otel_logger = Logger.new(STDOUT)
  OpenTelemetry.error_handler = lambda do |exception: nil, message: nil|
    updox_otel_logger.error "OpenTelemetry error: #{message}"
    updox_otel_logger.error exception
  end

  c.service_name = 'PartnerPortal'
  c.use 'OpenTelemetry::Instrumentation::ActionPack'
  c.use 'OpenTelemetry::Instrumentation::ActionView'
  c.use 'OpenTelemetry::Instrumentation::ActiveJob'
  c.use 'OpenTelemetry::Instrumentation::ActiveModelSerializers'
  c.use 'OpenTelemetry::Instrumentation::ActiveRecord'
  c.use 'OpenTelemetry::Instrumentation::ConcurrentRuby'
  c.use 'OpenTelemetry::Instrumentation::Dalli'
  c.use 'OpenTelemetry::Instrumentation::HttpClient'
  c.use 'OpenTelemetry::Instrumentation::Mysql2'
  c.use 'OpenTelemetry::Instrumentation::Rack'
  c.use 'OpenTelemetry::Instrumentation::Rails'
  c.use 'OpenTelemetry::Instrumentation::Redis'
  c.use 'OpenTelemetry::Instrumentation::Sidekiq'
  c.use 'OpenTelemetry::Instrumentation::Sinatra'
  # Avoiding use_all to isolate error messages
end
```

My `Gemfile` section for this is:

```ruby
# Telemetry
gem 'opentelemetry-sdk', '~> 1.0.1'
gem 'opentelemetry-exporter-otlp', '~> 0.20.6'
gem 'opentelemetry-instrumentation-all', '~> 0.21.3'
```

The server is started with the following script:

```bash
#!/bin/bash

# Run database migrations.
bin/rake db:migrate --trace RAILS_ENV=docker

echo ''
echo 'Finished DB migration. Starting server...'
echo ''

# Start server.
bundle exec rails server -e docker -b 0.0.0.0 -p ${PORT}
```

And the `OpenTracing` logs listed above are getting triggered by `bin/rake db:migrate ...` instead of the `bundle exec ...`

**Comment by arielvalentin** [comment, marked as answer]
👋🏼 @3ygun It looks like you are using a version of Rails (5.0.2) that is now [EOL](https://endoflife.date/rails). The minimum version OTEL Ruby Auto-Instrumentation will support is 5.2

Take a look at this appraisal file to see all supported versions:

https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/rails/Appraisals

**Reply by 3ygun** [reply]
@arielvalentin Thanks for the information on that. I didn't know there was an "Appraisals" file. I had looked for the minimal rails version on the https://rubygems.org/gems/opentelemetry-instrumentation-rails page which shows as `rails >= 0` which obviously I meet with an EOL version BUT not with the Appraisals file.

Either way that's an actionable next step for me to upgrade our rails version if possible.

![opentelemetry-instrumentation-rails-gem-page](https://user-images.githubusercontent.com/5482320/144510039-db389be8-a428-4de2-9055-71ac0a9e8d07.png)

**Reply by 3ygun** [reply]
Based on your and @robertlaurin 's answers the versions that I have present in the application are probably the issue and I'll open another discussion if that isn't the case.

**Reply by arielvalentin** [reply]
Thank you for the feedback. We need to update our documentation to make it easier for folks to reason about version compatibility but I agree with you that our gem specifications could be more helpful in this regard.  

cc: @ericmustin @robertlaurin  https://github.com/open-telemetry/opentelemetry-ruby/pull/1025

**Comment by robertlaurin** [comment]
Hey, 

Sharing some references to see if we can understand what is going wrong here together.

Those failure messages are from what I'll call "soft failures" 
https://github.com/open-telemetry/opentelemetry-ruby/blob/d397e1f3e60cfdcc2af2d7acca2d0f9cd104c55f/instrumentation/base/lib/opentelemetry/instrumentation/registry.rb#L77-L85

That means during the install process failed somewhere here
https://github.com/open-telemetry/opentelemetry-ruby/blob/d397e1f3e60cfdcc2af2d7acca2d0f9cd104c55f/instrumentation/base/lib/opentelemetry/instrumentation/base.rb#L209-L222

So using Rails as an example failure that means `defined?(::Rails)` evaluated to false *or* `gem_version >= MINIMUM_VERSION` evaluated to false
https://github.com/open-telemetry/opentelemetry-ruby/blob/d397e1f3e60cfdcc2af2d7acca2d0f9cd104c55f/instrumentation/rails/lib/opentelemetry/instrumentation/rails/instrumentation.rb#L14-L22

I see that you're using `Using rails 5.0.7.2` so that means it would fail to install.  The other gems are quite possibly failing for other similar checks.

**Reply by 3ygun** [reply]
Thanks for the detailed answer! Makes sense. Is there a reason that the error message is so basic even for something that re-portability ("gem version too old!") doesn't get logged to the user?

**Reply by robertlaurin** [reply]
There's no particular reason for it being so generic.  I suppose we just haven't had an ask for it yet.  I'll look at adding a bit more details for the various failure modes.  However because the `present` and `compatible` blocks accept arbitrary code it will still be at the coarse level of "failed due to compatibility check" or "failed due to present check".

**Comment by grvm** [comment]
As a workaround for evaluation purposes, this helped me install everything & get it working for Rails 5.1. Has been working properly since the last 24 hours.

Update your initializer to include the following

```ruby
module OpenTelemetry
  module Instrumentation
    Rails::Instrumentation::MINIMUM_VERSION = Gem::Version.new('5.1.0')
    ActiveRecord::Instrumentation::MINIMUM_VERSION = Gem::Version.new('5.1.0')
    ActionPack::Instrumentation::MINIMUM_VERSION = Gem::Version.new('5.1.0')
    ActionView::Instrumentation::MINIMUM_VERSION = Gem::Version.new('5.1.0')
  end
end
```

**Reply by ahayworth** [reply]
This will indeed allow the install to proceed - just be aware that we do not support those versions and make no guarantees about behavior or performance in this case. ❤️

**Reply by grvm** [reply]
Yup, learned it the hard way. Things completely broke, basically the `ActiveRecord` `save!` method

**Reply by ahayworth** [reply]
Oh no! I'm sorry to hear that, I was hoping for more success for you. ❤️
